common/xdc/eventReplication.go

```diff
@@ -0,0 +1,128 @@
+package xdc
+
+import (
+	"github.com/uber/cadence/common"
+)
+
+const (
+	emptyCommonAncestor = -1
+)
+type (
+	eventNode struct {
+		eventID int64
+		version int64
+	}
+
+	eventBranch struct {
+		branchToken []byte
+		versionHistory []eventNode
+	}
+
+	eventReplicationMetadata struct {
+		eventsBranches [] eventBranch
+	}
+)
+
+//Not thread-safe
+func newEventBranch(token []byte, eventNodes ...eventNode) *eventBranch {
+	return &eventBranch{
+		branchToken: token,
+		versionHistory: eventNodes,
+	}
+}
+
+// add is to add new node into the versionHistory slice
+func (e *eventBranch) add(node eventNode) {
+	if node.eventID <= e.getLastNode().eventID || node.version <= e.getLastNode().version {
+		// Or panic?
+		return
+	}
+	e.versionHistory = append(e.versionHistory, node)
+}
+
+// updateEventID updates the last event id
+func (e *eventBranch) updateEventID(eventID int64) {
+	if eventID <= e.getLastNode().eventID {
+		// Or panic?
+		return
+	}
+	e.getLastNode().eventID = eventID
+}
+
+// append is to add or update the versionHistory slice
+func (e *eventBranch) append(node eventNode) {
+	if node.version < e.getLastNode().version {
+		// Or panic?
+		return
+	}
+
+	if node.version > e.getLastNode().version {
+		e.add(node)
+	} else {
+		e.updateEventID(node.eventID)
+	}
+}
+
+func (e *eventBranch) getLastNode() eventNode {
+	return e.versionHistory[len(e.versionHistory)-1]
+}
+
+func newEventReplicationMetadata(branches ...eventBranch) *eventReplicationMetadata {
+	return &eventReplicationMetadata {
+		eventsBranches: branches,
+	}
+}
+
+func (m *eventReplicationMetadata) merge(branch eventBranch) {
+	eventID, branchIdx := m.findBranchWithMaxEventID(branch)
+
+	if m.eventsBranches[branchIdx].versionHistory[len(m.eventsBranches[branchIdx].versionHistory) - 1].eventID == eventID {
+		for _, event := range branch.versionHistory {
+			if event.eventID > eventID {
+				m.eventsBranches[branchIdx].append(event)
+			}
+		}
+	} else {
+		m.eventsBranches = append(m.eventsBranches, branch)
+	}
+}
+
+
+//Lowest common ancestor among all branches
+func (m *eventReplicationMetadata) findBranchWithMaxEventID(branch eventBranch) (eventID int64, branchIndex int) {
+	for i := 0; i < len(m.eventsBranches); i++ {
+		eID := m.findMinEventIDWithMaxSameVersion(m.eventsBranches[i], branch)
+		if eID < 0 {
+			return emptyCommonAncestor, emptyCommonAncestor
+		}
+		if eID > eventID {
+			eventID = eID
+			branchIndex = i
+		}
+	}
+	return
+}
+
+func (m *eventReplicationMetadata) findMinEventIDWithMaxSameVersion(local, remote eventBranch) int64 {
+	localIdx := len(local.versionHistory) - 1
```

Comment:
how about this
```
length := len(local.versionHistory)
if length > len(remote.versionHistory) {
	length = len(remote.versionHistory)
}

index := -1
for i := 0; i < length; i++ {
	if local.versionHistory[i].version == remote.versionHistory[i].version {
		index = i
	}
}

if index < 0 {
	panic("version history is malformed")
}

eventID := common.MinInt64(local.versionHistory[index].eventID, remote.versionHistory[index].eventID)
version := local.versionHistory[index].version
return eventNode{
	version: version,
	eventID: eventID,
}
```

Reply:
Which one do you prefer? This is top down search and mine is bottom up ?